File: Backend/src/main/java/lan/dk/podcastserver/manager/worker/updater/DailymotionUpdater.java
package lan.dk.podcastserver.manager.worker.updater;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.jayway.jsonpath.TypeRef;
import javaslang.collection.HashSet;
import javaslang.collection.Set;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.service.ImageService;
import lan.dk.podcastserver.service.JsonService;
import lan.dk.podcastserver.service.SignatureService;
import lan.dk.podcastserver.service.properties.PodcastServerParameters;
import lombok.Getter;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.springframework.stereotype.Component;

import javax.validation.Validator;
import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.regex.Pattern;

import static lan.dk.podcastserver.utils.MatcherExtractor.PatternExtractor;
import static lan.dk.podcastserver.utils.MatcherExtractor.from;

/**
 * Created by kevin on 21/02/2016 for Podcast Server
 */
@Slf4j
@Component("DailymotionUpdater")
public class DailymotionUpdater extends AbstractUpdater {

    static final String API_LIST_OF_ITEMS = "https://api.dailymotion.com/user/%s/videos?fields=created_time,description,id,thumbnail_720_url,title";
    // http://www.dailymotion.com/karimdebbache
    private static final PatternExtractor USER_NAME_EXTRACTOR = from(Pattern.compile("^.+dailymotion.com/(.*)"));
    private static final String ITEM_URL = "http://www.dailymotion.com/video/%s";
    private static final TypeRef<Set<DailymotionVideoDetail>> LIST_DAILYMOTIONVIDEODETAIL_TYPE = new TypeRef<Set<DailymotionVideoDetail>>() { };

    private final JsonService jsonService;
    private final ImageService imageService;

    public DailymotionUpdater(PodcastServerParameters podcastServerParameters, SignatureService signatureService, Validator validator, ImageService imageService, JsonService jsonService) {
        super(podcastServerParameters, signatureService, validator);
        this.imageService = imageService;
        this.jsonService = jsonService;
    }

    @Override
    public Set<Item> getItems(Podcast podcast) {
        return USER_NAME_EXTRACTOR.on(podcast.getUrl()).group(1)
                .map(username -> String.format(API_LIST_OF_ITEMS, username))
                .flatMap(jsonService::parseUrl)
                .map(p -> p.read("list", LIST_DAILYMOTIONVIDEODETAIL_TYPE))
                .map(this::asSet)
                .getOrElse(HashSet::empty);
    }

    private Set<Item> asSet(Set<DailymotionVideoDetail> jsonArray) {
        return jsonArray
                .map(i -> Item.builder()
                        .url(String.format(ITEM_URL, i.getId()))
                        .cover(imageService.getCoverFromURL(i.getCover()))
                        .title(i.getTitle())
                        .pubDate(ZonedDateTime.ofInstant(Instant.ofEpochSecond(i.getCreationDate()), ZoneId.of("Europe/Paris")))
                        .description((i.getDescription()))
                        .build()
                );
    }

    @Override
    public String signatureOf(Podcast podcast) {
        return USER_NAME_EXTRACTOR.on(podcast.getUrl()).group(1)
                .map(u -> signatureService.generateSignatureFromURL(String.format(API_LIST_OF_ITEMS, u)))
                .getOrElseThrow(() -> new RuntimeException("Username not Found"));
    }

    @Override
    public Type type() {
        return new AbstractUpdater.Type("Dailymotion", "Dailymotion");
    }

    @Override
    public Integer compatibility(String url) {
        return StringUtils.contains(url, "www.dailymotion.com") ? 1 : Integer.MAX_VALUE;
    }

    private static class DailymotionVideoDetail {
        @Getter @Setter private String id;
        @JsonProperty("thumbnail_720_url") @Getter @Setter private String cover;
        @Getter @Setter private String title;
        @JsonProperty("created_time") @Getter @Setter private Long creationDate;
        @Getter @Setter private String description;

    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/update/updaters/dailymotion/DailymotionUpdater.kt
package com.github.davinkevin.podcastserver.update.updaters.dailymotion

import com.fasterxml.jackson.annotation.JsonProperty
import com.github.davinkevin.podcastserver.service.image.ImageService
import com.github.davinkevin.podcastserver.update.fetchCoverUpdateInformation
import com.github.davinkevin.podcastserver.update.updaters.ItemFromUpdate
import com.github.davinkevin.podcastserver.update.updaters.PodcastToUpdate
import com.github.davinkevin.podcastserver.update.updaters.Type
import com.github.davinkevin.podcastserver.update.updaters.Updater
import com.github.davinkevin.podcastserver.utils.MatcherExtractor
import org.springframework.util.DigestUtils
import org.springframework.web.client.RestClient
import org.springframework.web.client.body
import java.net.URI
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime

class DailymotionUpdater(
    private val rc: RestClient,
    private val image: ImageService
): Updater {

    override fun signatureOf(url: URI): String? {
        val userName = USER_NAME_EXTRACTOR
            .on(url.toASCIIString())
            .group(1)
            ?: error("username not found")

        val channelDetails = rc
            .get()
            .uri("/user/{userName}/videos?fields=id", userName)
            .retrieve()
            .body<DailymotionResult>()
            ?: return null

        if (channelDetails.list.isEmpty()) {
            return ""
        }

        val sign = channelDetails.list
            .map { it.id }
            .sorted()
            .reduce { t, u -> """$t, $u""" }
            .let { DigestUtils.md5DigestAsHex(it.toByteArray()) }

        return sign
    }

    override fun findItems(podcast: PodcastToUpdate): List<ItemFromUpdate> {
        val userName = USER_NAME_EXTRACTOR.on(podcast.url.toASCIIString()).group(1) ?: error("username not found")

        val channelDetails = rc
            .get()
            .uri("/user/{userName}/videos?fields=created_time,description,id,thumbnail_720_url,title", userName)
            .retrieve()
            .body<DailymotionDetailsResult>()
            ?: return emptyList()

        return channelDetails.list.map {
            val cover = it.cover?.let(image::fetchCoverUpdateInformation)

            ItemFromUpdate(
                url = URI("https://www.dailymotion.com/video/${it.id}"),
                cover = cover,
                title = it.title,
                pubDate = ZonedDateTime.ofInstant(Instant.ofEpochSecond(it.creationDate!!), ZoneId.of("Europe/Paris")),
                description = it.description!!,
                mimeType = "video/mp4"
            )
        }
    }

    override fun type() = Type("Dailymotion", "Dailymotion")
    override fun compatibility(url: String): Int = when {
        "www.dailymotion.com" in url -> 1
        else -> Integer.MAX_VALUE
    }
}

// http://www.dailymotion.com/karimdebbache
private val USER_NAME_EXTRACTOR = MatcherExtractor.from("^.+dailymotion.com/(.*)")

private class DailymotionDetailsResult(val list: Set<DailymotionVideoDetail> = emptySet()) {
    class DailymotionVideoDetail(
        val id: String,
        val title: String,
        val description: String? = null,
        @JsonProperty("created_time") val creationDate: Long? = null,
        @JsonProperty("thumbnail_720_url") val cover: URI? = null
    )
}

private class DailymotionResult(val list: Set<DailymotionUpdaterVideoDetail> = emptySet()) {
    class DailymotionUpdaterVideoDetail(val id: String)
}
